packages/platform-ios/src/commands/runIOS/index.js

```diff
@@ -66,65 +71,36 @@ function runIOS(_: Array<string>, ctx: ConfigT, args: FlagsT) {
     }),
   );
 
-  if (args.device) {
-    const selectedDevice = matchingDevice(devices, args.device);
+  const device = ((args.device: any): string);
+  const udid = ((args.udid: any): string);
+  if (device || udid) {
```

Comment:
This is not clean, inside we need to branch 2 more times, which is not really more readable and efficient. What would you tell about this refactoring?

1. If there's no `device` nor `udid`, we run on simulator straight away.
1. If there are no devices found, we bail early.
1. We run device/udid branching inside `matchingDevice` fn (inlining `matchingDeviceByUdid` into it)
1. If device matches, run on device
1. If not, print appropriate messages depending on whether `device` or `udid` was passed.

```diff
--- a/packages/platform-ios/src/commands/runIOS/index.js
+++ b/packages/platform-ios/src/commands/runIOS/index.js
@@ -64,6 +64,12 @@ function runIOS(_: Array<string>, ctx: ConfigT, args: FlagsT) {
     }`,
   );

+  const {device, udid} = args;
+
+  if (!device && !udid) {
+    return runOnSimulator(xcodeProject, scheme, args);
+  }
+
   const devices = parseIOSDevicesList(
     // $FlowExpectedError https://github.com/facebook/flow/issues/5675
     child_process.execFileSync('xcrun', ['instruments', '-s'], {
@@ -71,33 +77,31 @@ function runIOS(_: Array<string>, ctx: ConfigT, args: FlagsT) {
     }),
   );

-  const device = ((args.device: any): string);
-  const udid = ((args.udid: any): string);
-  if (device || udid) {
-    const selectedDevice = device
-      ? matchingDevice(devices, device)
-      : matchingDeviceByUdid(devices, udid);
+  if (devices.length === 0) {
+    return logger.error('No iOS devices connected.');
+  }
+
+  const selectedDevice = matchingDevice(devices, device, udid);

   if (selectedDevice) {
     return runOnDevice(selectedDevice, scheme, xcodeProject, args);
   }

-    if (devices && devices.length > 0) {
-      const message = device
-        ? `Could not find device with the name: "${device}". Choose one of the following:\n${printFoundDevices(
-            devices,
-          )}`
-        : `Could not find device with the udid: "${udid}". Choose one of the following:\n${printFoundDevices(
-            devices,
-          )}`;
-
-      return logger.error(message);
+  if (device) {
+    return logger.error(
+      `Could not find device with the name: "${String(
+        device,
+      )}". Choose one of the following:\n${printFoundDevices(devices)}`,
+    );
   }

-    return logger.error('No iOS devices connected.');
+  if (udid) {
+    return logger.error(
+      `Could not find device with the udid: "${udid}". Choose one of the following:\n${printFoundDevices(
+        devices,
+      )}`,
+    );
   }
-
-  return runOnSimulator(xcodeProject, scheme, args);
 }

 async function runOnSimulator(xcodeProject, scheme, args: FlagsT) {
@@ -331,7 +335,10 @@ function xcprettyAvailable() {
   return true;
 }

-function matchingDevice(devices, deviceName) {
+function matchingDevice(devices, deviceName, udid) {
+  if (udid) {
+    matchingDeviceByUdid(devices, udid);
+  }

```

Reply:
I usually prefer to return as early as possible, even if it needs to have negative conditionals, with a quick look I agree with the diff you sent.

@grabbou thoughts on this? Asking because you pointed that you didn't like the conditionals I introduced on previous commits.

Reply:
I actually disagree this is not clean. The first branch is `runOnDevice` - you can select device by name or id. The second branch is `runOnSimulator`. 

Anyway, I don't have a strong preference here. Let's merge it as soon as possible as I want to ship it for next RC of React Native.

We can tweak it later.

Reply:
PS. I like @thymikee structure better, but at the same time, I feel like @lucasbento keeps doing tweaking that is far beyond his PR. I'd just get it merged to not discourage them from further PRs!

Reply:
@grabbou feel free to merge it as is now. Hopefully @lucasbento will not be discouraged by our nitpicks 😅